javascript/src/URDFClasses.js

```diff
@@ -87,7 +87,9 @@ class URDFJoint extends URDFBase {
                 break;
 
             case 'planar':
-                this.jointValue = new Array(2).fill(0);
+                // Planar joints rotate about their Z axis
+                this.axis = new Vector3(0, 0, 1);
+                this.jointValue = new Array(3).fill(0);
```

Comment:
I didn't realize planar joints could rotate? Is there documentation on this? [The specification](https://wiki.ros.org/urdf/XML/joint) is pretty sparse when it comes to the descriptions on joint behavior but it makes it sound like planar movement only? If the ROS implementation differs from the spec then it might be worth getting the wiki page updated.

> planar — this joint allows motion in a plane perpendicular to the axis.

Reply:
I'll need to dig into this more. I was pointed to some code in MoveIt and ROS Control that treats them as X,Y,Rz - I was also previously under the impression that it was only X,Y.

Reply:
Info seems pretty scattered but from [these urdfpy docs](https://urdfpy.readthedocs.io/en/latest/generated/urdfpy.Joint.html) it looks like a "planar" joint is described as a joint that only supports planar movement on x and y in the plane described by the axis vector. It's not clear how you're supposed to derive the local x and y axes from the axis vector, though.

Reply:
For context, here's [the code in MoveIt that clearly treats it as having a rotational component](https://github.com/ros-planning/moveit2/blob/f47b4c01b7508079925bb8ce733ea9bc22582127/moveit_core/robot_model/src/planar_joint_model.cpp#L345-L349).

But you're definitely right that that's not how it's implemented in `urdfpy`. I'm trying to dig into how this is implemented in RViz, because I assume whatever MoveIt is doing is visualized fine in RViz.

Reply:
I'll discuss this more internally with the MoveIt & MoveIt Pro team and determine whether we actually think this is part of the standard or whether it's some incorrect assumption made long ago. `git blame` says that code is 12 years old!

If it's a MoveIt-specific thing then I'll have the implementation here be 2dof and we'll fix it on the MoveIt end.

Reply:
I couldn't find a consistent definition of a planar joint. 
However, I came across a few other libraries used for manipulator motion planning that include the rotational component in their Planar joint definition.  

Matlab - https://www.mathworks.com/help/sm/ref/planarjoint.html
Drake - https://drake.mit.edu/doxygen_cxx/classdrake_1_1multibody_1_1_planar_joint.html
[Comsol](https://www.comsol.com/) - https://doc.comsol.com/5.5/doc/com.comsol.help.mbd/mbd_ug_modeling.3.07.html
MapleSim - https://www.maplesoft.com/support/help/maplesim/view.aspx?path=componentLibrary/multibody/jointsAndMotions

@gkjohnson Would it be okay to add the rotational component to the planar joint here? This would help with our current development without huge changes to MoveIt.

Reply:
What's your motivation here? The spec does not read so ambiguously to me and the URDF py project explicitly does not support a rotational axis. URDF is a defined _specification_ and I'm generally not keen on adding new features outside of the spec (for this or other formats) just because it's convenient. It inevitably leads to situations like this where users will produce files that are designed with a single, spec-breaking implementation in mind but which will not work in applications that have followed it, ultimately diluting the purpose of the spec.

If this is something that's important I recommend making suggestions to edit the URDF specification. I understand that ROS is already a bit of a wild west in terms of software but the least we can do is not continue to proliferate inconsistencies with the few formats that are in place to enable interoperability.

Reply:
Unfortunately every attempt to update the URDF specification that I'm aware of has stalled for years at a time :(

My understanding is that URDFpy is the outlier here and other URDF parsers (i.e. urdf-rs) do support a rotational axis.

Reply:
we'll reopen the conversation internally about removing the rotation from planar joints. If you're certain we don't want to support a rotation component here, I can remove it from this PR and move just that addition to another branch I'll use on my project for now.

Reply:
I'm mostly asking for a good faith effort to be made to improve the specification if it seems that it's out of date or not clearly representative of the common interpretation, as you are suggesting, so this kind of confusion does not happen in the future. It is not difficult to change the wiki page - I made a change myself years ago at afford [more geometry formats](https://wiki.ros.org/action/info/urdf/XML/link?action=diff&rev2=72&rev1=71) to be allowed in the specification. If all but urdf-py is interpreting planar joints as "two translation axes and one rotation axis" then the documentation should be updated to reflect that more clearly and at least an issue should be made for urdf-py to bring it in line with other tools. You can request edit access for your wiki account [here](https://github.com/ros-infrastructure/roswiki/issues/258).

> Unfortunately every attempt to update the URDF specification that I'm aware of has stalled for years at a time :(

If you're referring to [URDF 2.0](https://sachinchitta.github.io/urdf2/), then yeah it's been stalled forever and I'll be surprised if it goes anywhere because of how large and breaking the the scope is. But making clarifying, non breaking changes to the URDF spec is not the same as trying to replace it completely.

Reply:
thanks for the clarification. We'll start a conversation on the ROS wiki side!

This has been a huge can of worms across the board, honestly - we went to polish up mobile base support in our application and found issues all across the ROS stack (even to the point of straight-up segfaults when you pass a planar joint to things). As always, I appreciate your engagement with my issues and PRs here. That's more than can be said for a lot of the packages we use!